File: python/src/train/models/fat/ffn.py
"""Field-Aware Feed-Forward Network for FAT (Field-Aware Transformer).

Paper §3.3: Replaces the standard shared FFN with field-conditioned FFN:

    FAT-FFN(z_i) = SiLU(z_i · W_1^{(f_i)}) · W_2^{(f_i)}

Each field f has dedicated projection matrices W_1^{(f)} and W_2^{(f)},
synthesised by the Basis-Composed Hypernetwork.

Reference:
    https://arxiv.org/abs/2511.12081
"""

from __future__ import annotations

import torch
import torch.nn as nn
import torch.nn.functional as F


class FieldAwareFFN(nn.Module):
    """Field-conditioned FFN with per-field W₁, W₂ projections.

    Args:
        d: Model dimension.
        d_ff: FFN intermediate dimension.
    """

    def __init__(self, d: int, d_ff: int) -> None:
        super().__init__()
        self.d = d
        self.d_ff = d_ff

        # LayerNorm applied before FFN (Pre-LN style)
        self.norm = nn.LayerNorm(d)

    def forward(
        self,
        x: torch.Tensor,
        W_1: torch.Tensor,
        W_2: torch.Tensor,
    ) -> torch.Tensor:
        """Forward pass with pre-computed field-specific projections.

        Args:
            x: [batch, num_fields, d]  input token embeddings.
            W_1: [num_fields, d, d_ff]  field-specific W₁.
            W_2: [num_fields, d_ff, d]  field-specific W₂.

        Returns:
            output: [batch, num_fields, d]
        """
        # Pre-LN
        x = self.norm(x)

        # FAT-FFN: SiLU(x · W₁) · W₂
        #   x: [B, F, d],  W_1: [F, d, d_ff]  →  hidden: [B, F, d_ff]
        hidden = torch.einsum("bfd,fde->bfe", x, W_1)
        hidden = F.silu(hidden)
        #   hidden: [B, F, d_ff],  W_2: [F, d_ff, d]  →  output: [B, F, d]
        output = torch.einsum("bfe,fed->bfd", hidden, W_2)

        return output

File: src/models/fat/ffn.rs
//! Field-Aware FFN (§3.3) — per-field W₁ and W₂ projections with SiLU activation.
//!
//! FAT-FFN(z_i) = SiLU(z_i · W_1^{(f_i)}) · W_2^{(f_i)}

use candle_core::{Module, Result, Tensor};
use candle_nn::{layer_norm, LayerNorm, VarBuilder};

/// Field-Aware Feed-Forward Network.
#[allow(dead_code)]
pub struct FieldAwareFFN {
    d: usize,
    d_ff: usize,
    norm: LayerNorm,
}

impl FieldAwareFFN {
    /// Create a new field-aware FFN layer.
    pub fn new(vb: VarBuilder, d: usize, d_ff: usize) -> Result<Self> {
        let norm = layer_norm(d, 1e-5, vb.pp("norm"))?;
        Ok(Self { d, d_ff, norm })
    }

    /// Forward pass with pre-computed field-specific projections.
    ///
    /// Args:
    ///   x: [batch, num_fields, d]  input.
    ///   w_1: [num_fields, d, d_ff]  field-specific input projections.
    ///   w_2: [num_fields, d_ff, d]  field-specific output projections.
    ///
    /// Returns: [batch, num_fields, d]
    #[allow(non_snake_case)]
    pub fn forward(&self, x: &Tensor, w_1: &Tensor, w_2: &Tensor) -> Result<Tensor> {
        // Pre-LN
        let x = self.norm.forward(x)?;

        // FAT-FFN: SiLU(x · W₁) · W₂
        // SiLU(x) = x * sigmoid(x)  (identical to Swish with β=1)
        let hidden = self.field_matmul(&x, w_1)?;           // [B, F, d_ff]
        let sig = candle_nn::ops::sigmoid(&hidden)?;
        let hidden = hidden.mul(&sig)?;                     // SiLU activation
        let output = self.field_matmul(&hidden, w_2)?;     // [B, F, d]

        Ok(output)
    }

    /// Batched per-field matrix multiply:  out[b,f,:] = x[b,f,:] @ W[f,:,:]
    fn field_matmul(&self, x: &Tensor, w: &Tensor) -> Result<Tensor> {
        let (_batch, num_fields, _) = x.dims3()?;
        let mut outputs = Vec::with_capacity(num_fields);
        for f in 0..num_fields {
            let x_f = x.narrow(1, f, 1)?.squeeze(1)?;     // [B, d_in]
            let w_f = w.get(f)?.squeeze(0)?;                // [d_in, d_out]
            let out_f = x_f.matmul(&w_f)?;                  // [B, d_out]
            outputs.push(out_f.unsqueeze(1)?);               // [B, 1, d_out]
        }
        Tensor::cat(&outputs, 1)
    }
}
